nginx/nginx.go

```diff
@@ -0,0 +1,233 @@
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		d, err := kc.ExtensionsV1beta1().Deployments(n.Namespace).Get(n.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if d.Status.UpdatedReplicas != d.Status.AvailableReplicas && d.Status.UnavailableReplicas != 0 {
+			return fmt.Errorf("expected: %d ready: %d", d.Status.UpdatedReplicas, d.Status.AvailableReplicas)
+		}
+
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+
+	//wait for all pods to enter running phase
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		pl, err := kc.CoreV1().Pods(n.Namespace).List(metav1.ListOptions{
+			LabelSelector: metav1.FormatLabelSelector(n.labelSel),
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(pl.Items) == 0 {
+			return fmt.Errorf("there should be non-zero no. of pods")
+		}
+
+		var pods []*v1.Pod
+		for i := range pl.Items {
+			p := &pl.Items[i]
+			if p.Status.Phase != v1.PodRunning {
+				return fmt.Errorf("pod %s not running", p.GetName())
+			}
+
+			pods = append(pods, p)
+		}
+
+		n.Pods = pods
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+
+	//create nginx service
+	svc := n.getServiceSpecs()
+	if _, err := kc.CoreV1().Services(n.Namespace).Create(svc); err != nil {
+		return nil, fmt.Errorf("error creating service %s: %v", n.SVCName, err)
+	}
+	return n, nil
+}
+
+// Ping pings the nginx service.
+// runs `wget --timeout 5 <nginx-svc-name>`
+// expectedPhase is the phase you expect the ping pods to end up in.
+func (n *Nginx) Ping(expectedPhase v1.PodPhase) error {
+	pp := n.getPingPodSpecs()
+	if _, err := n.kc.CoreV1().Pods(n.Namespace).Create(pp); err != nil {
+		return err
+	}
+
+	// wait for pod state
+	if err := utils.Retry(5, 5*time.Second, func() error {
+		p, err := n.kc.CoreV1().Pods(n.Namespace).Get(pp.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if p.Status.Phase != expectedPhase {
+			return fmt.Errorf("expected phase: %v found: %v", expectedPhase, p.Status.Phase)
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete deletes the deployment and service
+func (n *Nginx) Delete() error {
+	delPropPolicy := metav1.DeletePropagationForeground
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		if err := n.kc.ExtensionsV1beta1().Deployments(n.Namespace).Delete(n.Name, &metav1.DeleteOptions{
+			PropagationPolicy: &delPropPolicy,
+		}); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		if err := n.kc.CoreV1().Services(n.Namespace).Delete(n.SVCName, &metav1.DeleteOptions{
+			PropagationPolicy: &delPropPolicy,
+		}); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (n *Nginx) getDeploymentSpecs() *extensionsv1beta1.Deployment {
```

Comment:
nit: `get` in method names not idiomatic (https://golang.org/doc/effective_go.html#Getters)